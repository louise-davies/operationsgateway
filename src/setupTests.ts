// jest-dom adds custom jest matchers for asserting on DOM nodes.
// allows you to do things like:
// expect(element).toHaveTextContent(/react/i)
// learn more: https://github.com/testing-library/jest-dom
import '@testing-library/jest-dom';
import crypto from 'crypto';
// TODO JOEL: Is this still needed? - Also in vite config
// need to mock <canvas> for plotting
import 'vitest-canvas-mock';
// TODO JOEL: Replace with vitest-fail-on-console
import failOnConsole from 'jest-fail-on-console';
import { TextEncoder } from 'util';
import { server } from './mocks/server';

// TODO JOEL: Is this still needed? - same for failOnConsole below
global.TextEncoder = TextEncoder;

failOnConsole();

vi.setConfig({ testTimeout: 15000 });

// Establish API mocking before all tests.
beforeAll(() => server.listen());

// Reset any request handlers that we may add during the tests,
// so they don't affect other tests.
afterEach(() => server.resetHandlers());

// Clean up after the tests are finished.
afterAll(() => server.close());

// TODO JOEL: Check if a all of the below is still needed

if (typeof window.URL.createObjectURL === 'undefined') {
  // required as work-around for enzyme/jest environment not implementing window.URL.createObjectURL method
  Object.defineProperty(window.URL, 'createObjectURL', {
    value: () => 'testObjectUrl',
  });
}

if (typeof window.URL.revokeObjectURL === 'undefined') {
  // required as work-around for enzyme/jest environment not implementing window.URL.createObjectURL method
  Object.defineProperty(window.URL, 'revokeObjectURL', {
    value: () => {},
  });
}

// this is needed because of https://github.com/facebook/jest/issues/8987
// eslint-disable-next-line @typescript-eslint/ban-ts-comment
// @ts-ignore
let mockActualReact;
vi.doMock('react', async () => {
  // eslint-disable-next-line @typescript-eslint/ban-ts-comment
  // @ts-ignore
  if (!mockActualReact) {
    mockActualReact = await vi.importActual('react');
  }
  return mockActualReact;
});

if (typeof window.URL.createObjectURL === 'undefined') {
  // required as work-around for enzyme/jest environment not implementing window.URL.createObjectURL method
  Object.defineProperty(window.URL, 'createObjectURL', {
    value: () => 'testObjectUrl',
  });
}

if (typeof window.URL.revokeObjectURL === 'undefined') {
  // required as work-around for enzyme/jest environment not implementing window.URL.revokeObjectURL method
  Object.defineProperty(window.URL, 'revokeObjectURL', {
    value: () => {
      // no-op
    },
  });
}

// jest doesn't implement web crypto so set up nodejs crypto as a default
Object.defineProperty(global, 'crypto', {
  value: Object.setPrototypeOf({ subtle: crypto.subtle }, crypto),
});
