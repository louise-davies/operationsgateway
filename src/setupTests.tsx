// jest-dom adds custom jest matchers for asserting on DOM nodes.
// allows you to do things like:
// expect(element).toHaveTextContent(/react/i)
// learn more: https://github.com/testing-library/jest-dom
import '@testing-library/jest-dom';
// need to mock <canvas> for plotting
import 'jest-canvas-mock';
import {
  FullChannelMetadata,
  PlotDataset,
  timeChannelName,
  FullScalarChannelMetadata,
  DEFAULT_WINDOW_VARS,
} from './app.types';
import { Action, PreloadedState, ThunkAction } from '@reduxjs/toolkit';
import { AppStore, RootState, setupStore } from './state/store';
import { initialState as initialConfigState } from './state/slices/configSlice';
import { initialState as initialTableState } from './state/slices/tableSlice';
import { initialState as initialSearchState } from './state/slices/searchSlice';
import {
  initialState as initialPlotState,
  PlotConfig,
} from './state/slices/plotSlice';
import { initialState as initialFilterState } from './state/slices/filterSlice';
import { render } from '@testing-library/react';
import type { RenderOptions } from '@testing-library/react';
import { Provider } from 'react-redux';
import { QueryClientProvider, QueryClient } from '@tanstack/react-query';
import { COLOUR_ORDER } from './plotting/plotSettings/colourGenerator';
import { staticChannels } from './api/channels';
import { server } from './mocks/server';
import { matchRequestUrl, MockedRequest } from 'msw';
import channelsJson from './mocks/channels.json';

jest.setTimeout(15000);

// Establish API mocking before all tests.
beforeAll(() => server.listen());

// Reset any request handlers that we may add during the tests,
// so they don't affect other tests.
afterEach(() => server.resetHandlers());

// Clean up after the tests are finished.
afterAll(() => server.close());

/**
 * Waits for msw request -
 * @param method string representing the HTTP method
 * @param url string representing the URL match for the route
 * @returns a promise of the matching request
 *  */
export function waitForRequest(method: string, url: string) {
  let requestId = '';

  return new Promise<MockedRequest>((resolve, reject) => {
    server.events.on('request:start', (req) => {
      const matchesMethod = req.method.toLowerCase() === method.toLowerCase();

      const matchesUrl = matchRequestUrl(req.url, url).matches;

      if (matchesMethod && matchesUrl) {
        requestId = req.id;
      }
    });

    server.events.on('request:match', (req) => {
      if (req.id === requestId) {
        resolve(req);
      }
    });

    server.events.on('request:unhandled', (req) => {
      if (req.id === requestId) {
        reject(
          new Error(`The ${req.method} ${req.url.href} request was unhandled.`)
        );
      }
    });
  });
}

// this is needed because of https://github.com/facebook/jest/issues/8987
// eslint-disable-next-line @typescript-eslint/ban-ts-comment
// @ts-ignore
let mockActualReact;
jest.doMock('react', () => {
  // eslint-disable-next-line @typescript-eslint/ban-ts-comment
  // @ts-ignore
  if (!mockActualReact) {
    mockActualReact = jest.requireActual('react');
  }
  return mockActualReact;
});

export let actions: Action[] = [];
export const resetActions = (): void => {
  actions = [];
};
export const getInitialState = (): RootState => ({
  config: initialConfigState,
  table: initialTableState,
  search: initialSearchState,
  plots: initialPlotState,
  filter: initialFilterState,
});
export const dispatch = (
  action: Action | ThunkAction<void, RootState, unknown, Action<string>>
): void | Promise<void> => {
  if (typeof action === 'function') {
    action(dispatch, getInitialState, null);
    return Promise.resolve();
  } else {
    actions.push(action);
  }
};

if (typeof window.URL.createObjectURL === 'undefined') {
  // required as work-around for enzyme/jest environment not implementing window.URL.createObjectURL method
  Object.defineProperty(window.URL, 'createObjectURL', {
    value: () => 'testObjectUrl',
  });
}

if (typeof window.URL.revokeObjectURL === 'undefined') {
  // required as work-around for enzyme/jest environment not implementing window.URL.revokeObjectURL method
  Object.defineProperty(window.URL, 'revokeObjectURL', {
    value: () => {
      // no-op
    },
  });
}

// This type interface extends the default options for render from RTL, as well
// as allows the user to specify other things such as initialState, store.
interface ExtendedRenderOptions extends Omit<RenderOptions, 'queries'> {
  preloadedState?: PreloadedState<RootState>;
  store?: AppStore;
  queryClient?: QueryClient;
}

export const createTestQueryClient = (): QueryClient =>
  new QueryClient({
    defaultOptions: {
      queries: {
        retry: false,
        staleTime: 300000,
      },
    },
    logger: {
      log: console.log,
      warn: console.warn,
      error: jest.fn(),
    },
  });

export const hooksWrapperWithProviders = (
  state = {},
  queryClient?: QueryClient
) => {
  const testQueryClient = queryClient ?? createTestQueryClient();
  const store = setupStore(state);
  const wrapper = ({ children }) => (
    <Provider store={store}>
      <QueryClientProvider client={testQueryClient}>
        {children}
      </QueryClientProvider>
    </Provider>
  );
  return wrapper;
};

export function renderComponentWithProviders(
  ui: React.ReactElement,
  {
    preloadedState = {},
    // Automatically create a store instance if no store was passed in
    store = setupStore(preloadedState),
    // Automatically create a query client instance if no query client was passed in
    queryClient = createTestQueryClient(),
    ...renderOptions
  }: ExtendedRenderOptions = {}
) {
  function Wrapper({
    children,
  }: React.PropsWithChildren<unknown>): JSX.Element {
    return (
      <Provider store={store}>
        <QueryClientProvider client={queryClient}>
          {children}
        </QueryClientProvider>
      </Provider>
    );
  }
  return {
    store,
    queryClient,
    ...render(ui, { wrapper: Wrapper, ...renderOptions }),
  };
}

export function renderComponentWithStore(
  ui: React.ReactElement,
  {
    preloadedState = {},
    // Automatically create a store instance if no store was passed in
    store = setupStore(preloadedState),
    ...renderOptions
  }: ExtendedRenderOptions = {}
) {
  function Wrapper({
    children,
  }: React.PropsWithChildren<unknown>): JSX.Element {
    return <Provider store={store}>{children}</Provider>;
  }
  return { store, ...render(ui, { wrapper: Wrapper, ...renderOptions }) };
}

export const flushPromises = (): Promise<void> =>
  new Promise((resolve) => setTimeout(resolve));

// MUI date pickers default to mobile versions during testing and so functions
// like .simulate('change') will not work, this workaround ensures desktop
// datepickers are used in tests instead
// https://github.com/mui/material-ui-pickers/issues/2073
export const applyDatePickerWorkaround = (): void => {
  // add window.matchMedia
  // this is necessary for the date picker to be rendered in desktop mode.
  // if this is not provided, the mobile mode is rendered, which might lead to unexpected behavior
  Object.defineProperty(window, 'matchMedia', {
    writable: true,
    value: (query: string) => ({
      media: query,
      // this is the media query that @material-ui/pickers uses to determine if a device is a desktop device
      matches: query === '(pointer: fine)',
      onchange: () => undefined,
      addEventListener: () => undefined,
      removeEventListener: () => undefined,
      addListener: () => undefined,
      removeListener: () => undefined,
      dispatchEvent: () => false,
    }),
  });
};

export const cleanupDatePickerWorkaround = (): void => {
  delete window.matchMedia;
};

export const testChannels = [
  ...Object.values(staticChannels),
  ...Object.entries(channelsJson.channels).map(
    ([systemName, channel]) =>
      ({
        systemName,
        ...channel,
      } as FullChannelMetadata)
  ),
];

export const testScalarChannels: FullScalarChannelMetadata[] = [
  ...Object.values(staticChannels),
  ...Object.entries(channelsJson.channels)
    .filter(([systemName, channel]) => channel.type === 'scalar')
    .map(
      ([systemName, channel]) =>
        ({
          systemName,
          ...channel,
        } as FullScalarChannelMetadata)
    ),
];

export const generatePlotDataset = (num: number) => {
  const datasetName = testScalarChannels[num].systemName;
  const plotDataset: PlotDataset = {
    name: datasetName,
    data: [
      {
        timestamp: num,
        [datasetName]: num + num,
      },
      {
        timestamp: num + num,
        [datasetName]: num + num + num,
      },
      {
        timestamp: num + num + num,
        [datasetName]: num + num + num + num,
      },
    ],
  };
  return plotDataset;
};

export const testPlotDatasets = Array.from(Array(3), (_, i) =>
  generatePlotDataset(i + 1)
);

export const generatePlotConfig = (num: number) => {
  const plotTitle = `Plot ${num}`;

  const plotConfig: PlotConfig = {
    open: num % 2 === 0,
    title: plotTitle,
    plotType: num % 2 === 0 ? 'scatter' : 'line',
    XAxis: num % 3 === 0 ? timeChannelName : undefined,
    XAxisScale:
      num % 3 === 0 ? 'time' : num % 3 === 1 ? 'linear' : 'logarithmic',
    selectedPlotChannels: [],
    leftYAxisScale: num % 2 === 0 ? 'linear' : 'logarithmic',
    rightYAxisScale: num % 2 === 0 ? 'logarithmic' : 'linear',
    gridVisible: num % 2 === 0,
    axesLabelsVisible: num % 2 !== 0,
    selectedColours: [],
    remainingColours: COLOUR_ORDER.map((colour) => colour),
    ...DEFAULT_WINDOW_VARS,
  };

  return plotConfig;
};

export const testPlotConfigs = Array.from(Array(3), (_, i) =>
  generatePlotConfig(i)
);
