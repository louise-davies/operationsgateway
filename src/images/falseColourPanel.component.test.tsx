import { fireEvent, screen, within } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import React from 'react';
import { flushPromises, renderComponentWithProviders } from '../setupTests';
import FalseColourPanel from './falseColourPanel.component';

describe('False colour panel component', () => {
  let props: React.ComponentProps<typeof FalseColourPanel>;
  const changeColourMap = jest.fn();
  const changeLowerLevel = jest.fn();
  const changeUpperLevel = jest.fn();

  beforeEach(() => {
    props = {
      colourMap: 'colourmap_1',
      lowerLevel: 0,
      upperLevel: 255,
      changeColourMap,
      changeLowerLevel,
      changeUpperLevel,
    };

    jest.clearAllMocks();
  });

  const createView = () => {
    return renderComponentWithProviders(<FalseColourPanel {...props} />);
  };

  it('renders correctly', async () => {
    const { asFragment } = createView();

    // "load" requests
    await flushPromises();

    expect(asFragment()).toMatchSnapshot();
  });

  it('renders colour map dropdown and changes colour map on change', async () => {
    const user = userEvent.setup();
    createView();

    // "load" requests
    await flushPromises();

    const select = screen.getByLabelText('Colour Map');
    await user.click(select);

    let dropdown = screen.getByRole('listbox', {
      name: 'Colour Map',
    });
    await user.click(
      within(dropdown).getByRole('option', { name: 'colourmap_2' })
    );

    expect(changeColourMap).toHaveBeenCalledWith('colourmap_2');

    jest.clearAllMocks();

    // check user can select "default" colourmap
    await user.click(select);
    dropdown = screen.getByRole('listbox', {
      name: 'Colour Map',
    });
    await user.click(within(dropdown).getByRole('option', { name: 'Default' }));

    expect(changeColourMap).toHaveBeenCalledWith(undefined);
  });

  it('renders lower level slider and changes lower level on change', async () => {
    createView();

    const sliderInput = screen.getByRole('slider', {
      name: 'Lower Level (LL)',
    });

    await fireEvent.change(sliderInput, { target: { value: 25 } });

    expect(changeLowerLevel).toHaveBeenCalledWith(25);
  });

  it('lower level slider cannot exceed upper level', async () => {
    createView();

    const upperSliderInput = screen.getByRole('slider', {
      name: 'Upper Level (UL)',
    });

    await fireEvent.change(upperSliderInput, { target: { value: 100 } });

    const lowerSliderInput = screen.getByRole('slider', {
      name: 'Lower Level (LL)',
    });

    await fireEvent.change(lowerSliderInput, { target: { value: 125 } });

    expect(changeLowerLevel).toHaveBeenCalledWith(100);
  });

  it('renders upper level slider and changes upper level on change', async () => {
    createView();

    const sliderInput = screen.getByRole('slider', {
      name: 'Upper Level (UL)',
    });

    await fireEvent.change(sliderInput, { target: { value: 50 } });

    expect(changeUpperLevel).toHaveBeenCalledWith(50);
  });

  it('upper level slider cannot exceed lower level', async () => {
    createView();

    const lowerSliderInput = screen.getByRole('slider', {
      name: 'Lower Level (LL)',
    });

    await fireEvent.change(lowerSliderInput, { target: { value: 100 } });

    const upperSliderInput = screen.getByRole('slider', {
      name: 'Upper Level (UL)',
    });

    await fireEvent.change(upperSliderInput, { target: { value: 75 } });

    expect(changeUpperLevel).toHaveBeenCalledWith(100);
  });

  it('can be disabled and re-enabled', async () => {
    const user = userEvent.setup();
    createView();

    // "load" requests
    await flushPromises();

    const falseColourSwitch = screen.getByRole('checkbox', {
      name: 'False Colour',
    });

    expect(falseColourSwitch).toBeChecked();

    await user.click(falseColourSwitch);

    expect(falseColourSwitch).not.toBeChecked();
    expect(changeColourMap).toHaveBeenCalledWith(undefined);
    expect(changeLowerLevel).toHaveBeenCalledWith(undefined);
    expect(changeUpperLevel).toHaveBeenCalledWith(undefined);

    jest.clearAllMocks();

    await user.click(falseColourSwitch);

    expect(falseColourSwitch).toBeChecked();
    expect(changeColourMap).toHaveBeenCalledWith(undefined);
    expect(changeLowerLevel).toHaveBeenCalledWith(0);
    expect(changeUpperLevel).toHaveBeenCalledWith(255);

    // check that it can restore a selected colourmap as well
    const select = screen.getByLabelText('Colour Map');
    await user.click(select);

    const dropdown = screen.getByRole('listbox', {
      name: 'Colour Map',
    });
    await user.click(
      within(dropdown).getByRole('option', { name: 'colourmap_2' })
    );

    await user.click(falseColourSwitch);
    jest.clearAllMocks();
    await user.click(falseColourSwitch);

    expect(falseColourSwitch).toBeChecked();
    expect(changeColourMap).toHaveBeenCalledWith('colourmap_2');
  });
});
