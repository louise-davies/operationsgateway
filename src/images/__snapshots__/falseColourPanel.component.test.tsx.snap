// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`False colour panel component renders correctly 1`] = `
<DocumentFragment>
  <div
    class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 css-1ps6pg7-MuiPaper-root"
  >
    <div
      class="css-d34ffb-MuiStack-root"
    >
      <div
        class="MuiFormGroup-root css-dmmspl-MuiFormGroup-root"
      >
        <label
          class="MuiFormControlLabel-root MuiFormControlLabel-labelPlacementEnd css-j204z7-MuiFormControlLabel-root"
        >
          <span
            class="MuiSwitch-root MuiSwitch-sizeMedium css-julti5-MuiSwitch-root"
          >
            <span
              class="MuiButtonBase-root MuiSwitch-switchBase MuiSwitch-colorPrimary Mui-checked PrivateSwitchBase-root MuiSwitch-switchBase MuiSwitch-colorPrimary Mui-checked Mui-checked css-5ryogn-MuiButtonBase-root-MuiSwitch-switchBase"
            >
              <input
                checked=""
                class="PrivateSwitchBase-input MuiSwitch-input css-1m9pwf3"
                type="checkbox"
              />
              <span
                class="MuiSwitch-thumb css-jsexje-MuiSwitch-thumb"
              />
              <span
                class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
              />
            </span>
            <span
              class="MuiSwitch-track css-1yjjitx-MuiSwitch-track"
            />
          </span>
          <span
            class="MuiTypography-root MuiTypography-body1 MuiFormControlLabel-label css-ahj2mt-MuiTypography-root"
          >
            False Colour
          </span>
        </label>
        <label
          class="MuiFormControlLabel-root Mui-disabled MuiFormControlLabel-labelPlacementEnd css-j204z7-MuiFormControlLabel-root"
        >
          <span
            class="MuiSwitch-root MuiSwitch-sizeMedium css-julti5-MuiSwitch-root"
          >
            <span
              aria-disabled="true"
              class="MuiButtonBase-root MuiSwitch-switchBase MuiSwitch-colorPrimary Mui-disabled Mui-disabled PrivateSwitchBase-root MuiSwitch-switchBase MuiSwitch-colorPrimary Mui-disabled Mui-disabled css-5ryogn-MuiButtonBase-root-MuiSwitch-switchBase"
              tabindex="-1"
            >
              <input
                class="PrivateSwitchBase-input MuiSwitch-input css-1m9pwf3"
                disabled=""
                type="checkbox"
              />
              <span
                class="MuiSwitch-thumb css-jsexje-MuiSwitch-thumb"
              />
            </span>
            <span
              class="MuiSwitch-track css-1yjjitx-MuiSwitch-track"
            />
          </span>
          <span
            class="MuiTypography-root MuiTypography-body1 MuiFormControlLabel-label Mui-disabled css-ahj2mt-MuiTypography-root"
          >
            Reverse Colour
          </span>
        </label>
        <label
          class="MuiFormControlLabel-root MuiFormControlLabel-labelPlacementEnd css-j204z7-MuiFormControlLabel-root"
        >
          <span
            class="MuiSwitch-root MuiSwitch-sizeMedium css-julti5-MuiSwitch-root"
          >
            <span
              class="MuiButtonBase-root MuiSwitch-switchBase MuiSwitch-colorPrimary PrivateSwitchBase-root MuiSwitch-switchBase MuiSwitch-colorPrimary css-5ryogn-MuiButtonBase-root-MuiSwitch-switchBase"
            >
              <input
                class="PrivateSwitchBase-input MuiSwitch-input css-1m9pwf3"
                type="checkbox"
              />
              <span
                class="MuiSwitch-thumb css-jsexje-MuiSwitch-thumb"
              />
              <span
                class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
              />
            </span>
            <span
              class="MuiSwitch-track css-1yjjitx-MuiSwitch-track"
            />
          </span>
          <span
            class="MuiTypography-root MuiTypography-body1 MuiFormControlLabel-label css-ahj2mt-MuiTypography-root"
          >
            Show extended colourmap options
          </span>
        </label>
      </div>
      <div
        class="MuiFormControl-root css-1nrlq1o-MuiFormControl-root"
      >
        <label
          class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined MuiFormLabel-colorPrimary MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined css-14s5rfu-MuiFormLabel-root-MuiInputLabel-root"
          data-shrink="false"
          id="colour-map-select-label"
        >
          Colour Map
        </label>
        <div
          class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-formControl  css-1yk1gt9-MuiInputBase-root-MuiOutlinedInput-root-MuiSelect-root"
        >
          <div
            aria-expanded="false"
            aria-haspopup="listbox"
            aria-labelledby="colour-map-select-label colour-map-select"
            class="MuiSelect-select MuiSelect-outlined MuiInputBase-input MuiOutlinedInput-input css-11u53oe-MuiSelect-select-MuiInputBase-input-MuiOutlinedInput-input"
            id="colour-map-select"
            role="button"
            tabindex="0"
          >
            <span
              class="notranslate"
            >
              â€‹
            </span>
          </div>
          <input
            aria-hidden="true"
            class="MuiSelect-nativeInput css-yf8vq0-MuiSelect-nativeInput"
            tabindex="-1"
            value=""
          />
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium MuiSelect-icon MuiSelect-iconOutlined css-hfutr2-MuiSvgIcon-root-MuiSelect-icon"
            data-testid="ArrowDropDownIcon"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M7 10l5 5 5-5z"
            />
          </svg>
          <fieldset
            aria-hidden="true"
            class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
          >
            <legend
              class="css-yjsfm1"
            >
              <span>
                Colour Map
              </span>
            </legend>
          </fieldset>
        </div>
      </div>
      <div
        class="MuiFormControl-root css-1nrlq1o-MuiFormControl-root"
      >
        <label
          class="MuiFormLabel-root MuiFormLabel-colorPrimary css-bmhlz8-MuiFormLabel-root"
          id="range-slider-label"
        >
          Level Range
        </label>
        <span
          class="MuiSlider-root MuiSlider-colorPrimary MuiSlider-sizeMedium MuiSlider-marked MuiSlider-trackFalse css-14rswqv-MuiSlider-root"
        >
          <span
            class="MuiSlider-rail css-14pt78w-MuiSlider-rail"
          />
          <span
            class="MuiSlider-track css-pm1ix7-MuiSlider-track"
            style="left: 0%; width: 100%;"
          />
          <span
            class="MuiSlider-mark MuiSlider-markActive css-6cwnna-MuiSlider-mark"
            data-index="0"
            style="left: 0%;"
          />
          <span
            aria-hidden="true"
            class="MuiSlider-markLabel MuiSlider-markLabelActive css-1eoe787-MuiSlider-markLabel"
            data-index="0"
            style="left: 0%;"
          >
            0
          </span>
          <span
            class="MuiSlider-mark css-1tfve6q-MuiSlider-mark"
            data-index="1"
            style="left: 12.156862745098039%;"
          />
          <span
            class="MuiSlider-mark css-1tfve6q-MuiSlider-mark"
            data-index="2"
            style="left: 24.705882352941178%;"
          />
          <span
            aria-hidden="true"
            class="MuiSlider-markLabel css-yafthl-MuiSlider-markLabel"
            data-index="2"
            style="left: 24.705882352941178%;"
          >
            63
          </span>
          <span
            class="MuiSlider-mark css-1tfve6q-MuiSlider-mark"
            data-index="3"
            style="left: 37.254901960784316%;"
          />
          <span
            class="MuiSlider-mark css-1tfve6q-MuiSlider-mark"
            data-index="4"
            style="left: 49.80392156862745%;"
          />
          <span
            aria-hidden="true"
            class="MuiSlider-markLabel css-yafthl-MuiSlider-markLabel"
            data-index="4"
            style="left: 49.80392156862745%;"
          >
            127
          </span>
          <span
            class="MuiSlider-mark css-1tfve6q-MuiSlider-mark"
            data-index="5"
            style="left: 62.35294117647059%;"
          />
          <span
            class="MuiSlider-mark css-1tfve6q-MuiSlider-mark"
            data-index="6"
            style="left: 74.90196078431373%;"
          />
          <span
            aria-hidden="true"
            class="MuiSlider-markLabel css-yafthl-MuiSlider-markLabel"
            data-index="6"
            style="left: 74.90196078431373%;"
          >
            191
          </span>
          <span
            class="MuiSlider-mark css-1tfve6q-MuiSlider-mark"
            data-index="7"
            style="left: 87.45098039215686%;"
          />
          <span
            class="MuiSlider-mark MuiSlider-markActive css-6cwnna-MuiSlider-mark"
            data-index="8"
            style="left: 100%;"
          />
          <span
            aria-hidden="true"
            class="MuiSlider-markLabel MuiSlider-markLabelActive css-1eoe787-MuiSlider-markLabel"
            data-index="8"
            style="left: 100%;"
          >
            255
          </span>
          <span
            class="MuiSlider-thumb MuiSlider-thumbColorPrimary MuiSlider-thumbSizeMedium css-eg0mwd-MuiSlider-thumb"
            data-focusvisible="false"
            data-index="0"
            style="left: 0%;"
          >
            <input
              aria-labelledby="range-slider-label"
              aria-orientation="horizontal"
              aria-valuemax="255"
              aria-valuemin="0"
              aria-valuenow="0"
              data-index="0"
              max="255"
              min="0"
              step="1"
              style="border: 0px; height: 100%; margin: -1px; overflow: hidden; padding: 0px; position: absolute; white-space: nowrap; width: 100%; direction: ltr;"
              type="range"
              value="0"
            />
            <span
              aria-hidden="true"
              class="MuiSlider-valueLabel css-nnid7-MuiSlider-valueLabel"
            >
              <span
                class="MuiSlider-valueLabelCircle"
              >
                <span
                  class="MuiSlider-valueLabelLabel"
                >
                  0
                </span>
              </span>
            </span>
          </span>
          <span
            class="MuiSlider-thumb MuiSlider-thumbColorPrimary MuiSlider-thumbSizeMedium css-eg0mwd-MuiSlider-thumb"
            data-focusvisible="false"
            data-index="1"
            style="left: 100%;"
          >
            <input
              aria-labelledby="range-slider-label"
              aria-orientation="horizontal"
              aria-valuemax="255"
              aria-valuemin="0"
              aria-valuenow="255"
              data-index="1"
              max="255"
              min="0"
              step="1"
              style="border: 0px; height: 100%; margin: -1px; overflow: hidden; padding: 0px; position: absolute; white-space: nowrap; width: 100%; direction: ltr;"
              type="range"
              value="255"
            />
            <span
              aria-hidden="true"
              class="MuiSlider-valueLabel css-nnid7-MuiSlider-valueLabel"
            >
              <span
                class="MuiSlider-valueLabelCircle"
              >
                <span
                  class="MuiSlider-valueLabelLabel"
                >
                  255
                </span>
              </span>
            </span>
          </span>
        </span>
      </div>
      <img
        alt="Colour bar"
      />
    </div>
  </div>
</DocumentFragment>
`;
